File: src/identify.py
import os
import google.generativeai as genai
import json

# Fetching the API key from environment
api_key = os.environ.get("GEMINI_API_KEY") 
if not api_key:
    # Look for API key around the common configuration places.
    import pathlib
    print("WARNING: Make sure you have GEMINI_API_KEY set. Exiting.")
    exit(1)

genai.configure(api_key=api_key)

dir_path = "C:/Users/zackc/Pictures/Sigma Camera Pictures/Sigma Flora"
files = [f for f in os.listdir(dir_path) if f.endswith(".jpg")]

results = {}

for file in files:
    try:
        print(f"Processing {file}...")
        img = genai.upload_file(os.path.join(dir_path, file))
        model = genai.GenerativeModel("gemini-1.5-pro")
        
        prompt = "Identify the flower in this photo. " + \
                 "Provide the scientific name. " + \
                 "Also provide a VERY concise description of what is happening in the photo (max 5-6 words), e.g., 'flower name (scientific name) in fence'. " + \
                 "Return exactly two lines. Line 1: Common Name. Line 2: The short description."
                 
        start_prompt = "Return exactly two lines. Line 1: Common Name. Line 2: Short Description."
        response = model.generate_content([img, prompt])
        lines = response.text.strip().split('\n')
        
        results[file] = {
            "name": lines[0].strip(),
            "description": lines[1].strip() if len(lines) > 1 else ""
        }
    except Exception as e:
        print(f"Error on {file}: {e}")

with open("flora_identifications.json", "w") as f:
    json.dump(results, f, indent=2)

print("Done.")

File: src/identify.js
const { GoogleGenerativeAI } = require("@google/generative-ai");
const fs = require("fs");
const path = require("path");

const apiKey = process.env.GEMINI_API_KEY;
if (!apiKey) {
    console.error("WARNING: Make sure you have GEMINI_API_KEY set. Exiting.");
    process.exit(1);
}

const genAI = new GoogleGenerativeAI(apiKey);
const dirPath = "C:/Users/zackc/Pictures/Sigma Camera Pictures/Sigma Flora";

function fileToGenerativePart(filePath, mimeType) {
    return {
        inlineData: {
            data: Buffer.from(fs.readFileSync(filePath)).toString("base64"),
            mimeType
        },
    };
}

async function run() {
    const files = fs.readdirSync(dirPath).filter(f => f.endsWith(".jpg"));
    const results = {};

    const model = genAI.getGenerativeModel({ model: "gemini-1.5-flash" });

    for (const file of files) {
        console.log(`Processing ${file}...`);
        try {
            const prompt = "Identify the flower in this photo. Provide the common name (if any) and the scientific name. Also provide a VERY concise description of what is happening in the photo (max 5-6 words), e.g., 'flower name (scientific name) in fence'. Return exactly two lines. Line 1: Common Name (Scientific Name). Line 2: The short description.";

            const imagePart = fileToGenerativePart(path.join(dirPath, file), "image/jpeg");

            const result = await model.generateContent([prompt, imagePart]);
            const response = await result.response;
            const text = response.text();

            const lines = text.trim().split("\n").map(l => l.trim()).filter(l => l);

            results[file] = {
                name: lines[0] || "",
                description: lines[1] || ""
            };
        } catch (e) {
            console.error(`Error on ${file}:`, e);
        }
    }

    fs.writeFileSync("flora_identifications.json", JSON.stringify(results, null, 2));
    console.log("Done.");
}

run();
